### PR TITLE
feat: add extra resources support

### LINKED ISSUE #77: Support requesting extra resources
### What problem are you facing?

We manage infrastructure with crossplane. Our compositions refer to some observe only resources (via the new beta style management policies). As of today, we need to create observe-only resources for each composition. Those resources could be global networks, flavors/hardware templates, or other resources that are commonly provided by the cloud provider.

### How could this Function help solve your problem?

With the new feature of requesting extra resources in Crossplane 1.15, go template could fetch those extra resources and provide them as variables in the go template via the RunFunction request. This would have the following benefits:

- Less observed resources needed (1 per XRD/Composition, instead of 1 per XR) -> therefore, there is less load on the provider
- Faster composition creations. We do not need to wait until the observe resources are ready.

### PR DESCRIPTION
**Pull request by phisco**
### Description of your changes

Fixes #77
 
Testing example:
```sh
$ go run ./... --debug --insecure &
$ cd example/extra-resources/
# add back the annotations to run the local 
#   annotations:
#     render.crossplane.io/runtime: "Development"
#     render.crossplane.io/runtime-development-target: "localhost:9443"
$ crossplane beta render xr.yaml composition.yaml functions.yaml --extra-resources extraResources.yaml -r
---
apiVersion: example.crossplane.io/v1beta1
kind: XR
metadata:
  name: example
status:
  conditions:
  - lastTransitionTime: "2024-01-01T00:00:00Z"
    message: 'Unready resources: bucket-configmap-0'
    reason: Creating
    status: "False"
    type: Ready
  dummy: cool-status
---
apiVersion: kubernetes.crossplane.io/v1alpha1
kind: Object
metadata:
  annotations:
    crossplane.io/composition-resource-name: bucket-configmap-0
  generateName: example-
  labels:
    crossplane.io/composite: example
  ownerReferences:
  - apiVersion: example.crossplane.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: XR
    name: example
    uid: ""
spec:
  forProvider:
    manifest:
      apiVersion: v1
      data:
        bucket: random-bucket-id
      kind: Configmap
      metadata:
        name: my-awesome-dev-bucket-bucket
  providerConfigRef:
    name: kubernetes
```

Once this pr is merged and a release is cut, renovate will bump the image pointed by the example.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit tests for my change.

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

**Comment by danielloader** [comment]
> misses an example, I'll try to find a meaningful one.

If you want an example the ability to provision an s3 bucket external to the collision (in a composition of its own or not) and be able to add bucket notifications from another composition would be a good way to demonstrate a one to many relationship. 

Often one bucket will have a dozen notification handlers into SQS queues etc.

**Comment by danielloader** [comment]
Comments from my day of using this branch:

- Document or change the nested "items" key on each extraResources object.
- Documentation around the lazy evaluation and guarding needed for the composition function in general, it's a wider composition function issue but it really bites when you're reaching out for additional resources and can't make use of the sequencer function. Especially when you're grabbing the external resource to grab some connection secret strings off it and the sole use it to combine it into the composition connection secrets (grabbing MongoDB Atlas hostname for example).

Ideas to track, if they should become issues or not is up to you:

- A function or documented functionality of how to guard the MRs from all the externalResource requests not being fulfilled at the templating time - causing malformed or missing fields in MRs to be templated.

Otherwise it's worked as advertised and enabled the workflows I hoped it would!

**Comment by dariozachow** [comment]
Is it planned to merge this one soon?

**Comment by markussiebert** [comment]
Can you provide a full working example? I can't get this running ...

**Comment by markussiebert** [comment]
My problem is, that I am able to retrieve extraResources with the function, but it's not possible to use them as described:

```
cannot compose resources: pipeline step "go-templating" returned a fatal result: cannot execute template: template: manifests:10:26: executing "manifests" at <index .extraResources "namespaces">
```

with this template:
```
apiVersion: 'meta.gotemplating.fn.crossplane.io/v1alpha1'
kind: 'ExtraResources'
requirements:
  namespaces:
    apiVersion: 'kubernetes.crossplane.io/v1alpha2'
    kind: 'Object'
    matchLabels:
      'something/claimRefNamespace': 'string'
---
{{ $someExtraResources := index .extraResources "namespaces" }}
{{- range $i, $extraResource := $someExtraResources.items }}
apiVersion: kubernetes.crossplane.io/v1alpha1
kind: Object
metadata:
  annotations:
    gotemplating.fn.crossplane.io/composition-resource-name: default-serviceaccount-prod
spec:
  forProvider:
    manifest:
      apiVersion: v1
      kind: Configmap
      metadata:
        name: $extraResource.resource.metadata.name
        namespace: default
      data:
        somekey: value
  providerConfigRef:
    name: "kubernetes"
{{- end }}
```

If I surround the part after the extraResources with {{ with .extraResources }} I get the error message from crossplane, that the resource didn't stabilize.
So how is this ment to be used?

**Comment by take-five** [comment]
This seems super useful! Will this be merged soon?

File: fn_test.go [hunk 1]
```diff
@@ -587,6 +592,124 @@ func TestRunFunction(t *testing.T) {
 				},
 			},
 		},
+		"ExtraResources": {
+			reason: "The Function should return a fatal result if the extra resource key is duplicated.",
```
Comment (markussiebert) [inline fn_test.go:596]: wrong reason?

**Comment by markussiebert** [comment]
Sorry - have the question, does this PR has any future?

**Comment by phisco** [comment]
Yes, @markussiebert, I only have to find the time to refine it a bit further.

File: README.md [hunk 2]
```diff
@@ -106,6 +107,65 @@ $ crossplane beta render xr.yaml composition.yaml functions.yaml
...
+
+```yaml
+apiVersion: meta.gotemplating.fn.crossplane.io/v1alpha1
+kind: ExtraResources
+requirements:
+  some-foo-by-name:
+    # Resources can be requested either by name
+    apiVersion: example.com/v1beta1
+    kind: Foo
+    matchName: "some-extra-foo"
```
Comment (henrysachs) [inline README.md:124]: Hey There, I just tried to match a Namespaced Resource by name but it appears this isnt possible for namespaced resources. Is this assumption right? 

This is the corresponding error: 

projectname is a variable i'm trying to assign the result to

```
        cannot compose resources: fetching resources for projectName: cannot get
        extra resource by name: an empty namespace may not be set when a
        resource name is provided
```

this is my extra resource:

```
apiVersion: 'meta.gotemplating.fn.crossplane.io/v1alpha1'
kind: 'ExtraResources'
requirements:
  projectName:
    apiVersion: example.com/v1alpha1
    kind: ExampleProject
    matchName: {{ $spec.projectName }}
```

Comment (phisco) [reply README.md:124]: hi, yes, only cluster scoped resources are supported

File: extraresources.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	MatchName string `json:"matchName,omitempty"`
+}
+
+// ToResourceSelector converts the ExtraResourcesRequirement to a fnv1beta1.ResourceSelector.
+func (e *ExtraResourcesRequirement) ToResourceSelector() *fnv1beta1.ResourceSelector {
+	out := &fnv1beta1.ResourceSelector{
+		ApiVersion: e.APIVersion,
+		Kind:       e.Kind,
+	}
+	if len(e.MatchLabels) == 0 {
```
Comment (Avarei) [inline extraresources.go:32]: when trying to retrieve ALL resources of a kind I would usually specify an empty label Selector.
Maybe this should check for empty string in MatchName instead?

**Comment by djmcgreal-cc** [comment]
Hi @phisco, @Avarei. It'd be great if this could get over the line, because the ecosystem doesn't have a function that can get parameterised resources https://github.com/crossplane-contrib/function-extra-resources/issues/18. Any chance you'll have the time?

**Review by ezgidemirel — APPROVED** [review]
Thanks for the change @phisco LGTM!

**Comment by godspeed529** [comment]
This looks great!  When can we get a new release?

**Comment by Moulick** [comment]
So extraResource will still only support looking up cluster scoped objects, not namespaced. For namespaced resource, we continue to use the `kubernetes-provider` to wrap it in a observer policy object?

**Comment by jbw976** [comment]
> When can we get a new release?

This has been included in the v0.6.0 release: https://github.com/crossplane-contrib/function-go-templating/releases/tag/v0.6.0

**Comment by phisco** [comment]
> So extraResource will still only support looking up cluster scoped objects, not namespaced. For namespaced resource, we continue to use the `kubernetes-provider` to wrap it in a observer policy object?

Yes, @Moulick.

**Pull request merged**